File: tcp-examples/VisualBasic/VisualBasic/Form1.vb
Imports System

Public Class Form1
    Private WithEvents client As New Verisurf.AsyncSocket

    Shared client_message As String = ""
    Shared server_message As String = ""

    Private Sub Command1_LinkClicked(sender As Object, e As LinkLabelLinkClickedEventArgs) Handles Command1.LinkClicked
        If client.Connected() = False Then
            Try
                client = New Verisurf.AsyncSocket
                client.Connect("127.0.0.1", 33666)
                client_message = "Verisurf Connection Established"
            Catch ex As Exception
            End Try
        Else
            client_message = "Verisurf Connection Already Established"
            UpdateResults()
        End If
    End Sub

    Private Sub Command2_LinkClicked(sender As Object, e As LinkLabelLinkClickedEventArgs) Handles Command2.LinkClicked
        SendCommand("<command_list />")
    End Sub

    Private Sub Command3_LinkClicked(sender As Object, e As LinkLabelLinkClickedEventArgs) Handles Command3.LinkClicked
        If client.Connected() = True Then
            Dim openFileDialog1 As New OpenFileDialog()
            openFileDialog1.Filter = "Verisurf Files (*.mcam)|*.*"
            If openFileDialog1.ShowDialog() = System.Windows.Forms.DialogResult.OK Then
                SendCommand("<file_open filename='" + openFileDialog1.FileName + "' />")
            End If
        Else
            client_message = "Verisurf Connection Not Established"
            UpdateResults()
        End If
    End Sub

    Private Sub Command4_LinkClicked(sender As Object, e As LinkLabelLinkClickedEventArgs) Handles Command4.LinkClicked
        SendCommand("<inspect_plan_list />")
    End Sub

    Private Sub Command5_LinkClicked(sender As Object, e As LinkLabelLinkClickedEventArgs) Handles Command5.LinkClicked
        SendCommand("<inspect_plan_load id='0' />")
    End Sub

    Private Sub Command6_LinkClicked(sender As Object, e As LinkLabelLinkClickedEventArgs) Handles Command6.LinkClicked
        SendCommand("<inspect_plan_info />")
    End Sub

    Private Sub Command7_LinkClicked(sender As Object, e As LinkLabelLinkClickedEventArgs) Handles Command7.LinkClicked
        SendCommand("<inspect_object_measure id='0' />")
    End Sub

    Private Sub Command8_LinkClicked(sender As Object, e As LinkLabelLinkClickedEventArgs) Handles Command8.LinkClicked
        SendCommand("<inspect_object_info id='0' />")
    End Sub

    Private Sub Command9_LinkClicked(sender As Object, e As LinkLabelLinkClickedEventArgs) Handles Command9.LinkClicked
        If client.Connected() = True Then
            Dim saveFileDialog1 As New SaveFileDialog()
            saveFileDialog1.Filter = "Verisurf Files (*.mcam)|*.*"
            If saveFileDialog1.ShowDialog() = System.Windows.Forms.DialogResult.OK Then
                SendCommand("<file_save filename='" + saveFileDialog1.FileName + "' />")
            End If
        Else
            client_message = "Verisurf Connection Not Established"
            UpdateResults()
        End If
    End Sub

    Private Sub Command10_LinkClicked(sender As Object, e As LinkLabelLinkClickedEventArgs) Handles Command10.LinkClicked
        If client.Connected() = True Then
            client.Close()
            client_message = "Verisurf Connection Closed"
            UpdateResults()
        Else
            client_message = "Verisurf Connection Not Established"
            UpdateResults()
        End If
    End Sub

    Private Sub Button1_Click(sender As Object, e As EventArgs) Handles Button1.Click
        SendCommand(Command.Text)
    End Sub

    Sub UpdateResults()
        If client_message.Length() & server_message.Length() Then
            Dim message As String = ""
            If client_message.Length() Then
                message += "CLIENT: " + client_message & vbCrLf & vbCrLf
            End If
            If server_message.Length() Then
                message += "SERVER: " + server_message & vbCrLf & vbCrLf
            End If

            client_message = ""
            server_message = ""

            If InvokeRequired Then
                Invoke(Sub() Results.Text = message)
            Else
                Results.Text = message
            End If
        End If
    End Sub

    Private Sub SendCommand(ByVal command As String)
        If client.Connected() = True Then
            client_message = command
            client.Send(command & vbLf)
        Else
            client_message = "Verisurf Connection Not Established"
            UpdateResults()
        End If
    End Sub

    Private Sub socketDataArrival(ByVal SocketID As String, ByVal SocketData As String) Handles client.socketDataArrival
        server_message = SocketData
        UpdateResults()
    End Sub

    Private Sub socketFailed(ByVal SocketID As String) Handles client.socketFailed
        client_message = "Verisurf Connection Failed"
        UpdateResults()
    End Sub

    Private Sub Form1_Load(ByVal sender As Object, ByVal e As System.EventArgs) Handles MyBase.Load
        Dim ctrl As Control
        For Each ctrl In Me.Controls
            If (ctrl.GetType() Is GetType(LinkLabel)) Then
                Dim ll As LinkLabel = CType(ctrl, LinkLabel)
                ll.LinkBehavior = System.Windows.Forms.LinkBehavior.NeverUnderline
            End If
        Next
    End Sub
End Class

File: tcp-examples/CSharp/CSharp/Form1.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace CSharp
{
    public partial class Form1 : Form
    {
        private Verisurf.AsyncSocket client = new Verisurf.AsyncSocket();
        private string client_message = "";
        private string server_message = "";

        private void Command1_LinkClicked(object sender, LinkLabelLinkClickedEventArgs e)
        {
            if(client.Connected() == false)
            {
                try
                {
                    client = new Verisurf.AsyncSocket();
                    client.m_failed += new Verisurf.socketFailed(socketFailed);
                    client.m_dataarrival += new Verisurf.socketDataArrival(socketDataArrival);
                    client.Connect("127.0.0.1", 33666);
                    client_message = "Verisurf Connection Established";
                }
                catch (Exception ex)
                {
                    Console.WriteLine(ex);
                }
            }
            else
            {
                client_message = "Verisurf Connection Already Established";
                UpdateResults();
            }
        }

        private void Command2_LinkClicked(object sender, LinkLabelLinkClickedEventArgs e)
        {
            SendCommand("<command_list />");
        }

        private void Command3_LinkClicked(object sender, LinkLabelLinkClickedEventArgs e)
        {
            if (client.Connected())
            {
                OpenFileDialog openFileDialog1 = new OpenFileDialog();
                openFileDialog1.Filter = "Verisurf Files (*.mcam)|*.*";
                if (openFileDialog1.ShowDialog() == System.Windows.Forms.DialogResult.OK)
                {
                    SendCommand("<file_open filename='" + openFileDialog1.FileName + "' />");
                }
            }
            else
            {
                client_message = "Verisurf Connection Not Established";
                UpdateResults();
            }
        }

        private void Command4_LinkClicked(object sender, LinkLabelLinkClickedEventArgs e)
        {
            SendCommand("<inspect_plan_list />");
        }

        private void Command5_LinkClicked(object sender, LinkLabelLinkClickedEventArgs e)
        {
            SendCommand("<inspect_plan_load id='0' />");
        }

        private void Command6_LinkClicked(object sender, LinkLabelLinkClickedEventArgs e)
        {
            SendCommand("<inspect_plan_info />");
        }

        private void Command7_LinkClicked(object sender, LinkLabelLinkClickedEventArgs e)
        {
            SendCommand("<inspect_object_measure id='0' />");
        }

        private void Command8_LinkClicked(object sender, LinkLabelLinkClickedEventArgs e)
        {
            SendCommand("<inspect_object_info id='0' />");
        }

        private void Command9_LinkClicked(object sender, LinkLabelLinkClickedEventArgs e)
        {
            if (client.Connected())
            {
                SaveFileDialog saveFileDialog1 = new SaveFileDialog();
                saveFileDialog1.Filter = "Verisurf Files (*.mcam)|*.*";
                if (saveFileDialog1.ShowDialog() == System.Windows.Forms.DialogResult.OK)
                {
                    SendCommand("<file_save filename='" + saveFileDialog1.FileName + "' />");
                }
            }
            else
            {
                client_message = "Verisurf Connection Not Established";
                UpdateResults();
            }
        }

        private void Command10_LinkClicked(object sender, LinkLabelLinkClickedEventArgs e)
        {
            if (client.Connected())
            {
                client.Close();
                client_message = "Verisurf Connection Closed";
                UpdateResults();
            }
            else
            {
                client_message = "Verisurf Connection Not Established";
                UpdateResults();
            }
        }

        private void Button1_Click(object sender, EventArgs e)
        {
            SendCommand(Command.Text);
        }

        private void UpdateResults()
        {
            if(client_message.Length > 0 || server_message.Length > 0)
            {
                string message = "";
                if (client_message.Length > 0)
                    message += "CLIENT: " + client_message + "\r\n\r\n";
                if (server_message.Length > 0)
                    message += "SERVER: " + server_message + "\r\n\r\n";

                client_message = "";
                server_message = "";

                if(Results.InvokeRequired)
                {
                    Results.Invoke((MethodInvoker) delegate { Results.Text = message; });
                }
                else
                {
                    Results.Text = message;
                }
            }
        }

        private void SendCommand(string command)
        {
            if(client.Connected())
            {
                client_message = command;
                client.Send(command + "\n");
            }
            else 
            {
                client_message = "Verisurf Connection Not Established";
                UpdateResults();
            }
        }

        public void socketFailed(string SocketID)
        {
            client_message = "Verisurf Connection Failed";
            UpdateResults();
        }

        public void socketDataArrival(string SocketID, string SocketData)
        {
            server_message = SocketData;
            UpdateResults();
        }

        public Form1()
        {
            InitializeComponent();

            foreach (Control ctrl in this.Controls)
            {
                if (ctrl is LinkLabel)
                {
                    LinkLabel ll = (LinkLabel)ctrl;
                    ll.LinkBehavior = System.Windows.Forms.LinkBehavior.NeverUnderline;
                }
            }
        }
    }
}
